README.md
# hvvclient
A live departure board for HVV transport (busses, S-Bahn, subway etc.) in Hamburg, Germany, implemented via REST calls to the GeoFox API GTI

## Getting Started

### Prerequisites

You'll need a webserver with PHP support, plus php-xml and php-curl-module, for Ubuntu:
```
sudo apt-get install php-xml php-curl
```
Since the project uses the Geofox Thin Interface (GTI), which requires registration, you'll have to apply for API access, see [HVV page](https://www.hvv.de/de/fahrplaene/abruf-fahrplaninfos/datenabruf).

Please note that the Geofox API documentation etc. is in German, as well as the software output, because the entire story takes place in Hamburg.

### Installing

Install the ```assets``` and ```inc``` folders plus ```index.php``` to some directory on your machine where your web server can find it. Fill in the access credentials for the Geofox API in ```inc/hvvc_vars.php``` with your data, start up the web server (if it's not running already), and point your browser to the ```index.php```. 
Now you should see the live departures of all public transport running from the ```$stations``` array defined in ```hvvc_vars.php```, which you can - and should - change later on, of course.

## Testing

The ```devdata``` directory contains sample XML files that demonstrate the communication of the REST interface.
The "request"-data is produced by the hvvclient, whereas the "response" data is delivered by the Geofox server.
The ```DLResponse.xml``` provided contains realtime data for transport that are on time, delayed, or in traffic jam, as well as cancelled journeys and plan data where no realtime data is available.

The project is developed using the [NetBeans IDE](https://netbeans.apache.org/) and contains the NetBeans project data in its ```nbproject``` folder.

## Authors

* **axaneco** - *Initial work*

## Acknowledgements

Thanks to Simon for proofreading this README.

## License

This project is licensed under the GNU GPL V3 License - see the [LICENSE.md](LICENSE.md) file for details.


ReleaseNotes.md
# hvvclient Release Notes

## C2107.1

 * changed Icon Service URL to cloud as requested by HBT/geofox

## C2102.1

* reflect the changes in XML delivered by geofox (see 2102*xml data in devdata dir) regarding live info: live info criteria is now if a <delay> tag is delivered. If yes, we have live data, if no, it's timetable data.
* added POST/XML debugging features (and a debug switch) as network sniffing makes no longer sense due to https
* added version string, updated user agent string

## C2009.1

* switch to new Geofox GTI server

## C1911.1

* handle departures in the past
* added "dd" URL parameter to display delay in minutes

## C1904.4

* GeoFox URL production switch
* added "from" and "to" URL parameters

## C1904.2

* generalized departure and via stations 
* output as table option
* GeoFox URL now variable

## Initial upload

* initial version, API calls are working


index.php
<?php

/*
 * hvvclient
 * implements checkName and departureList calls to the the GEOFOX Thin Interface (GTI)
 * a Passenger Information System for the Hamburger Verkehrsverbund (HVV)
 * for details see sections 2.2 and 2.4 of GTI Handbuch V35.1 
 * https://gti.geofox.de/html/GTIHandbuch_p.html
 * 
 * @author axaneco
 * 
 */

include ('inc/hvvc_vars.php'); // vars + station query xml
include ('inc/hvvc_functions.php'); //functions
// parameter "from" given in URL?
if ($_GET["from"]) {
    $both_dirs = TRUE;
    $stations["dep"][0] = $_GET["from"];
}
// parameter "to" given in URL?
if ($_GET["to"]) {
    $both_dirs = FALSE;
    $stations["via"][0] = $_GET["to"];
}

// parameter "dd" given in URL?
if ($_GET["dd"]) {
    $ddelay = $_GET["dd"];
}

// get actual station IDs from GeoFox
$stat = get_station_keys($gfurl, $stations, $username, $password);

// create gti:departureList request
if ($both_dirs) { // both directions
    $dl_xml = create_gti_DLRequest($stat["dep"][0], $stat["dep"][1], $refday, $reftime, $maxlist, $maxtimeoffset);
} else {
    $dl_xml = create_gti_DLRequest($stat["dep"][0], $stat["dep"][1], $refday, $reftime, $maxlist, $maxtimeoffset, $stat["via"][1]);
}

// get departure list for dep
$res = call_gti_api($gfurl, 'departureList', $dl_xml, $username, $password);

// read result as xml
$requestxml = simplexml_load_string($dl_xml);
$resultxml = simplexml_load_string($res);

// Debugging XML
if ($hvvc_debug) {
    echo "<p style=\"font-family:'Courier New'\">";
    echo "XML start<br>";
    echo "Request XML:<br>";
    echo "<pre>";
    echo htmlentities($requestxml->asXML());
    echo "</pre></p>";
    echo "<p style=\"font-family:'Courier New'\">";
    echo "Response XML:<br>";
    echo "<pre>";
    $dom = new DOMDocument('1.0', 'utf-8');
    $dom->formatOutput = true;
    $xml = $resultxml->asXML();
    $dom->loadXML($xml);
    $xml_pretty = $dom->saveXML();
    echo htmlentities($xml_pretty);
    echo "</pre>";
    echo "<p style=\"font-family:'Courier New'\">";
    echo "XML end</p>";
}

echo "<span style='font-family:sans-serif;'>";
echo "<span style='font-size:16px;'>\n";

// write results
echo "<a href='https://geofox.hvv.de/jsf/home.seam' target='_blank'>";
echo "<img src='https://www.hvv.de/images/logo_hvv_110x25.png' alt='Mit dem HVV zu uns' height='25' border='0'/>";
echo "</a>\n";

// print departure list
if ($both_dirs) { // both directions
    echo "<br><br>Nächste Abfahrten ab " . $stat["dep"][0] . " (ungefiltert):<br><br>\n";
} else {
    echo "<br><br>Nächste Abfahrten ab " . $stat["dep"][0] . " Richtung " . $stat["via"][0] . ":<br><br>\n";
}
print_departures($resultxml, $maxlist, TRUE, $ddelay); // boole parameter: table display yes/no

echo "</span></span>\n";
?>

inc/hvvc_vars.php
<?php

/*
 * hvvclient
 * implements checkName and departureList calls to the the GEOFOX Thin Interface (GTI)
 * a Passenger Information System for the Hamburger Verkehrsverbund (HVV)
 * for details see sections 2.2 and 2.4 of GTI Handbuch V35.1 
 * https://gti.geofox.de/html/GTIHandbuch_p.html
 * 
 * @author axaneco
 * 
 */

// Version string
$hvvc_version = 'C2102.1A';
// debugging mode
$hvvc_debug = FALSE;
// geofox API url
$gfurl = 'https://gti.geofox.de/gti/public/';

// stations array
// first station: departures
// second station: via filter for direction control
// XXX filled by get_station_keys() with HVV station codes
$stations = array(  "dep" => array( "Hamburg Hbf", "XXX"), 
                    "via" => array( "", "XXX") 
);

// date
$refday = date("d.m.Y");
$reftime = date("H:i");

//display paramater
$maxlist = 30; // max list items for query results
$maxtimeoffset = 720; // results for max 12 hours in future

// replace this by your data or...
$username = 'username';
$password = 'password';

// ...include the data from an extra file
include 'credentials.php';

?>

inc/hvvc_functions.php
<?php

/*
 * hvvclient
 * implements checkName and departureList calls to the the GEOFOX Thin Interface (GTI)
 * a Passenger Information System for the Hamburger Verkehrsverbund (HVV)
 * for details see sections 2.2 and 2.4 of GTI Handbuch V35.1 
 * https://gti.geofox.de/html/GTIHandbuch_p.html
 * 
 * @author axaneco
 * 
 */
// get all the station keys
function get_station_keys($gfurl, $stations, $username, $password) { // stations, username and password come from the vars.php
    for ($x = 0, $keys = array_keys($stations), $c = count($keys); $x < $c; $x++) {
        if ($test_flag) {
            echo($keys[$x] . ": " . $stations[$keys[$x]][0] . $stations[$keys[$x]][1] . "<br>");
        }
        // create the gti:checkName request
        $cn_xml = create_gti_CNRequest($stations[$keys[$x]][0]);
        // call the gti:checkName API function
        $res = call_gti_api($gfurl, 'checkName', $cn_xml, $username, $password);
        // set station key in stations array
        $resultxml = simplexml_load_string($res);
        $stations[$keys[$x]][1] = $resultxml->results->id;
    }
    return $stations;
}

// create xml gti:CNRequest (checkName xml body)
function create_gti_CNRequest($stname) {
    // xml header
    $dom = new DOMDocument('1.0', 'utf-8');
    $dom->xmlStandalone = TRUE;
    $dom->formatOutput = TRUE;

    //set request type with attributes
    $root = $dom->createElement("gti:CNRequest");
    $root->setAttribute("xmlns:gti", "http://www.geofox.de/schema/geofoxThinInterface");

    // define structure and set parameters
    $dom->appendChild($root);
    $root->appendChild($n_theName = $dom->createElement("theName"));
    $n_theName->appendChild($n_name = $dom->createElement("name", $stname));
    $n_theName->appendChild($n_type = $dom->createElement("type", "STATION"));
    $root->appendChild($n_maxList = $dom->createElement("maxList", "5"));
    $root->appendChild($n_coordinateType = $dom->createElement("coordinateType", "EPSG_4326"));

    return $dom->saveXML();
}

// create xml gti:DLRequest (departureList xml body)
function create_gti_DLRequest($stname, $stid, $refday, $reftime, $maxlist, $maxtimeoffset, $filterid = FALSE) {
    // xml header
    $dom = new DOMDocument('1.0', 'utf-8');
    $dom->xmlStandalone = TRUE;
    $dom->formatOutput = TRUE;

    //set request type with attributes
    $root = $dom->createElement("gti:DLRequest");
    $root->setAttribute("xmlns:gti", "http://www.geofox.de/schema/geofoxThinInterface");

    // define structure and set parameters
    $dom->appendChild($root);
    //$root->appendChild($n_version = $dom->createElement("version", "35"));
    $root->appendChild($n_version = $dom->createElement("version", "38"));
    $root->appendChild($n_station = $dom->createElement("station"));
    $n_station->appendChild($n_name = $dom->createElement("name", $stname));
    $n_station->appendChild($n_city = $dom->createElement("city", "Hamburg"));
    $n_station->appendChild($n_id = $dom->createElement("id", $stid));
    $n_station->appendChild($n_type = $dom->createElement("type", "STATION"));
    $root->appendChild($n_time = $dom->createElement("time"));
    $n_time->appendChild($n_date = $dom->createElement("date", $refday));
    $n_time->appendChild($n_stime = $dom->createElement("time", $reftime));
    $root->appendChild($n_maxlist = $dom->createElement("maxList", $maxlist));
    if ($filterid) {
        $root->appendChild($n_filter = $dom->createElement("filter"));
        $n_filter->appendChild($n_stationids = $dom->createElement("stationIDs", $filterid));
    }
    $root->appendChild($n_maxtimeoffset = $dom->createElement("maxTimeOffset", $maxtimeoffset));
    $root->appendChild($n_userealtime = $dom->createElement("useRealtime", "true"));

    return $dom->saveXML();
}

// Call GTI API via cURL
function call_gti_api($gfurl, $gfunc, $http_body, $username, $password) { // gfunc here either checkName or departureList
    global $hvvc_debug;
    global $hvvc_version;
    // sign the api request
    $bin_signature = hash_hmac("sha1", $http_body, $password, true);
    $signature = base64_encode($bin_signature);
    // make UUID for X-TraceId
    $traceid = v4();
    // create API call URI
    $ch = curl_init($gfurl . $gfunc);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLINFO_HEADER_OUT, true);
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $http_body);
    // Set HTTP Header for POST request 
    curl_setopt($ch, CURLOPT_HTTPHEADER, array(
        'Accept: application/xml',
        'Content-Type: application/xml;charset=UTF-8',
        'geofox-auth-type: HmacSHA1',
        'geofox-auth-user: ' . $username,
        'geofox-auth-signature: ' . $signature,
        'User-Agent: ax-hvvclient ' . $hvvc_version,
        'X-Platform: web',
        'X-TraceId: ' . $traceid,
        'Content-Length: ' . strlen($http_body))
    );
    // Submit the POST request
    $resultxml = curl_exec($ch);
    if ($hvvc_debug) {
        echo "<p style=\"font-family:'Courier New'\">cURL start<br>";
        $cURL = curl_getinfo ($ch,  CURLINFO_HEADER_OUT); 
        //print_r($cURL);
        echo nl2br($cURL);
        echo "<br>cURL end</p>";
    }
    curl_close($ch);
    return $resultxml;
}

// tableswitch
function tab($table, $alignment = FALSE) {
    if ($table) {
        echo "</td>";
        if ($alignment) {
            echo "<td align=$alignment>";
        } else {
            echo "<td>";
        }
    }
}

// check disturbances
function check_disturbances($resultxml, $i, $tdelay, $table) {
    //$rt = $resultxml->departures[$i]->attributes->types[0]; // is either REALTIME or missing
    $rt = $resultxml->departures[$i]->delay; // is either 0, >0 or missing
    $tj = $resultxml->departures[$i]->attributes->types[1]; // is either ACCURATE or TRAFFIC_JAM
    $dis = FALSE; // no disturbance initially

    if ($table) {
        echo "<td>";
    }
    // no live info -> blank symbol
    if ($rt[0] === NULL) {
        echo "<img src='assets/images/empty.png' height='14' border='0'/>";
        $dis = TRUE;
    }
    // traffic jam -> black symbol
    if ($tj == 'TRAFFIC_JAM') {
        echo "<img src='assets/images/black.png' height='14' border='0'/>";
        $dis = TRUE;
    }
    // delay (without traffic jam) -> yellow symbol
    if ($tdelay > 0 && $dis == FALSE) {
        echo "<img src='assets/images/yellow.png' height='14' border='0'/>";
        $dis = TRUE;
    }
    $res = array("rt" => $rt, "dis" => $dis);
    return $res;
}

// sofort switch
function now($tdep, $table) {
    if ($tdep < 0) {
        echo ' ist abgefahren'; // bypass time garbage with departures in the past
    }
    if ($tdep == 0) {
        echo ' sofort';
    } else {
        if ($table) {
            echo ' &nbsp;&nbsp ';
        } else {
            echo ' in ';
        }
        echo $tdep . ' Minute';
        if ($tdep > 1) {
            echo 'n';
        } // minuten for 0, 2 - inf
    }
}

// print out the departure list
function print_departures($resultxml, $maxlist, $table = FALSE, $ddelay = FALSE) { // resultxml delivered by the GeoFox API, here: call_gti_api($gfunc, $http_body, $username, $password)
    if ($table) {
        echo "<table>\n";
    }
    if ($resultxml->returnCode == 'OK') {
        for ($i = 0; $i < $maxlist; $i++) {
            $id = $resultxml->departures[$i]->line->id;         // get bus id
            if ($id) { // go on only if there's a result in the xml 
                if ($table) {
                    echo "<tr>";
                }
                $toffset = $resultxml->departures[$i]->timeOffset;  // departure time offset in minutes from query 
                $tdelay = round(($resultxml->departures[$i]->delay) / 60, 0, PHP_ROUND_HALF_UP);  // planned/known delay, if any, converted to minutes
                $tdep = $toffset + $tdelay; // estimated departure time including known delay
                $ex = $resultxml->departures[$i]->extra; // extra trip
                $no = $resultxml->departures[$i]->cancelled; // trip cancelled
                // check (and display as icon) disturbances
                $dst = check_disturbances($resultxml, $i, $tdelay, $table);
                // live info and everything ok -> green icon
                //if ($dst["rt"] == 'REALTIME' && $dst["dis"] == FALSE) {
                if ($dst["rt"] == '0' && $dst["dis"] == FALSE) {
                    echo "<img src='assets/images/green.png' height='14' border='0'/>";
                }
                tab($table, 'center');
                echo "<img src='https://cloud.geofox.de/icon/line?height=14&amp;lineKey=" . $id . "'> ";   // line icon
                tab($table);
                if ($no) {
                    echo '<s>';
                } // strike if no journey
                echo $resultxml->departures[$i]->line->direction . ' '; // line direction 
                if ($no) {
                    echo '</s>';
                }
                tab($table, 'right');
                if ($no) {
                    echo '<s>';
                } // strike if no journey
                now($tdep, $table); // "sofort" switch
                if ($tdelay > 0 && $ddelay) { // print delay in minutes if present and display switch is on
                    echo ' (+' . $tdelay . ')';
                }
                if ($no) {
                    echo '</s>';
                }
                tab($table);
                if ($no) {
                    echo '<font color="red"> FÄLLT AUS</font>';
                }
                if ($ex) {
                    echo ' (Verstärkerfahrt)';
                }
                if ($table) {
                    echo "</td></tr>\n";
                } else {
                    echo "<br>\n";
                }
            }
        }
        echo "</table>\n";
    } else {
        echo 'Fehler: GeoFox returned an error';
    }
}

/**
 * 
 * Generate v4 UUID
 * 
 * Version 4 UUIDs are pseudo-random.
 * 
 * @author Andrew Moore
 * @link http://www.php.net/manual/en/function.uniqid.php#94959
 * 
 * source https://github.com/macx/rfc-4122-uuid/blob/master/src/uuid.php
 * 
 */
function v4() {
    return sprintf('%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
            // 32 bits for "time_low"
            mt_rand(0, 0xffff), mt_rand(0, 0xffff),
            // 16 bits for "time_mid"
            mt_rand(0, 0xffff),
            // 16 bits for "time_hi_and_version",
            // four most significant bits holds version number 4
            mt_rand(0, 0x0fff) | 0x4000,
            // 16 bits, 8 bits for "clk_seq_hi_res",
            // 8 bits for "clk_seq_low",
            // two most significant bits holds zero and one for variant DCE1.1
            mt_rand(0, 0x3fff) | 0x8000,
            // 48 bits for "node"
            mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
    );
}
